packages/tpetra/core/src/Tpetra_Details_extractBlockDiagonal.hpp

```diff
@@ -0,0 +1,106 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+#define TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+
+#include "TpetraCore_config.h"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Teuchos_RCP.hpp"
+
+/// \file Tpetra_Details_extractBlockDiagonal.hpp
+/// \brief Functions that allow for the extraction of a block diagonal
+/// from a Tpetra::CrsMatrix.
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra.  The file itself or its contents may disappear or
+///   change at any time.
+
+namespace Tpetra {
+namespace Details {
+
+
+template<class SparseMatrixType,
+         class MultiVectorType>
+void extractBlockDiagonal(const SparseMatrixType& A, MultiVectorType & diagonal) {
+  using local_map_type = typename SparseMatrixType::map_type::local_map_type;
+  using LO             = typename SparseMatrixType::local_ordinal_type;
+  using KCRS           = typename SparseMatrixType::local_matrix_type;
+  using lno_view_t     = typename KCRS::StaticCrsGraphType::row_map_type::const_type;
+  using lno_nnz_view_t = typename KCRS::StaticCrsGraphType::entries_type::const_type;
+  using scalar_view_t  = typename KCRS::values_type::const_type;
+  using local_mv_type  = typename MultiVectorType::dual_view_type::t_dev;
+  using range_type     = Kokkos::RangePolicy<typename SparseMatrixType::node_type::execution_space, size_t>;
```

Comment:
1. This means that you intend to run on device, if `execution_space` is `Cuda`.  See below.
2. The most appropriate index type is `LO`, not `size_t`, because you're iterating over local rows.

Reply:
Fixed